src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage_WriteBatching.cs

```diff
@@ -116,54 +117,58 @@ internal partial class SQLitePersistentStorage
                 // to proceed.
                 taskCompletionSource?.TrySetResult(0);
             }
+        }
 
-            return;
-
-            // Local functions
-            async Task<(Task previousTask, TaskCompletionSource<int> taskCompletionSource)> GetWriteTaskAsync()
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", OftenCompletesSynchronously = true)]
+        private static async ValueTask<(Task previousTask, TaskCompletionSource<int> taskCompletionSource)> GetWriteTaskAsync<TKey>(
+            SemaphoreSlim gate,
+            MultiDictionary<TKey, Action<SqlConnection>> keyToWriteActions,
+            Dictionary<TKey, Task> keyToWriteTask,
+            TKey key,
+            ArrayBuilder<Action<SqlConnection>> writesToProcess,
+            CancellationToken cancellationToken)
+        {
+            // Have to acquire the semaphore.  We're going to mutate the shared 'keyToWriteActions'
+            // and 'keyToWriteTask' collections.
+            //
+            // Note: by blocking on _writeQueueGate we are guaranteed to see all the writes
+            // performed by FlushAllPendingWritesAsync.
+            using (await gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
this can be an instance method now, right?  And it would be totally ok to reference _writeQueueGate.

Up to you if you think this is better.

Reply:
Missed that part. Fixed now.